src/core/Carousel/Carousel.tsx

```diff
@@ -0,0 +1,298 @@
+    setSlideCount(React.Children.count(props.children));
+  }, [props.children, setSlideCount]);
+
+  const items = React.useMemo(
+    () =>
+      React.Children.map(props.children, (child, index) =>
+        React.isValidElement(child)
+          ? React.cloneElement(child, { 'aria-hidden': index !== currentIndex })
+          : child,
+      ) ?? [],
+    [props.children, currentIndex],
+  );
+
+  return (
+    <ol
+      {...props}
+      className={cx('iui-carousel-slider', props.className)}
+      style={{
+        // height: '100%',
+        transform: `translateX(-${currentIndex * 100}%)`,
+        ...props.style,
+      }}
+    >
+      {items}
+    </ol>
+  );
+};
+
+const CarouselSlide = (props: React.ComponentPropsWithoutRef<'li'>) => {
+  const { className, children, ...rest } = props;
+
+  return (
+    <li
+      className={cx('iui-carousel-slider-item', className)}
+      role='tabpanel'
+      aria-roledescription='slide'
+      {...rest}
+    >
+      {children}
+    </li>
+  );
+};
+
+const CarouselNavigation = (props: React.ComponentPropsWithoutRef<'nav'>) => {
+  const { className, children, ...rest } = props;
+
+  const context = React.useContext(CarouselContext);
+  if (!context) {
+    throw new Error('TODO');
+  }
+
+  const { slideCount, setCurrentIndex, keysPressed } = context;
+
+  return (
+    <nav className={cx('iui-carousel-navigation', className)} {...rest}>
+      <div className='iui-carousel-navigation-left'>
+        <IconButton
+          styleType='borderless'
+          size='small'
+          tabIndex={-1}
+          onClick={() =>
+            setCurrentIndex((old) => (slideCount + old - 1) % slideCount)
+          }
+          data-pressed={keysPressed['ArrowLeft'] || undefined}
+        >
+          <SvgChevronLeft />
+        </IconButton>
+      </div>
+      {children}
+      <div className='iui-carousel-navigation-right'>
+        <IconButton
+          styleType='borderless'
+          size='small'
+          tabIndex={-1}
+          onClick={() =>
+            setCurrentIndex((old) => (slideCount + old + 1) % slideCount)
+          }
+          data-pressed={keysPressed['ArrowRight'] || undefined}
+        >
+          <SvgChevronRight />
+        </IconButton>
+      </div>
+    </nav>
+  );
+};
+
+const CarouselDots = (
+  props: {
+    length?: number;
+    maxCount?: number;
+  } & React.ComponentPropsWithoutRef<'ol'>,
+) => {
+  const { maxCount = 5, className, children, ...rest } = props;
+  const context = React.useContext(CarouselContext);
+  if (!context) {
+    throw new Error('TODO: this needs to be usable without context');
+  }
+
+  const { slideCount, currentIndex, setCurrentIndex } = context;
+
+  useTheme();
+
+  const dots = React.useMemo(
+    () =>
+      Array(slideCount)
+        .fill(null)
+        .map((_, index) => {
+          const firstDotIndex = Math.max(
+            0,
+            Math.min(
+              currentIndex - Math.ceil(maxCount / 2) + 1,
```

Comment:
lost me in maths :D

Reply:
it's just trying to calculate how many dots to show on each side.
- `maxCount / 2` gives "half" of the dots
- `Math.ceil`/`floor` handles odd numbers
- `+1`/`-1` is accounting for current dot
- `Math.max`/`Math.min` is for clamping.